sciris/sc_parallel.py

```diff
@@ -0,0 +1,260 @@
+
+    Version: 2018nov01
+     '''
+    # Set up processes to start asynchronously
+    if maxload  is None: maxload = 0.8
+    if interval is None: interval = 0.5
+    if maxtime  is None: maxtime = 36000
+    if label    is None: label = ''
+    else: label += ': '
+    if index is None:
+        pause = np.random.rand()*interval
+        index = ''
+    else:              
+        pause = index*interval
+    if maxload>1: maxload/100. # If it's >1, assume it was given as a percent
+    time.sleep(pause) # Give it time to asynchronize
+    
+    # Loop until load is OK
+    toohigh = True # Assume too high
+    count = 0
+    maxcount = maxtime/float(interval)
+    while toohigh and count<maxcount:
+        count += 1
+        currentload = psutil.cpu_percent(interval=0.1)/100. # If interval is too small, can give very inaccurate readings
+        if currentload>maxload:
+            if verbose: print(label+'CPU load too high (%0.2f/%0.2f); process %s queued %i times' % (currentload, maxload, index, count))
+            time.sleep(interval*2*np.random.rand()) # Sleeps for an average of refresh seconds, but do it randomly so you don't get locking
+        else: 
+            toohigh = False 
+            if verbose: print(label+'CPU load fine (%0.2f/%0.2f), starting process %s after %i tries' % (currentload, maxload, index, count))
+    return None
+
+
+
+class TaskArgs(ut.prettyobj):
+        ''' A class to hold the arguments for the task -- must match both parallelize() and parallel_task() '''
+        def __init__(self, func, index, iterval, itervar, args, kwargs, maxload):
+            self.func = func
+            self.index = index
+            self.iterval = iterval
+            self.itervar = itervar
+            self.args = args
+            self.kwargs = kwargs
+            self.maxload = maxload
+            return None
+
+
+
+def parallelize(func=None, iterarg=None, itervar=None, args=None, kwargs=None, ncpus=None, maxload=None):
+    '''
+    Shortcut for parallelizing a function. iterarg can be an iterable or an integer;
+    if the latter, it will run the function that number of times.
+    
+    The function can either handle a fixed number of CPUs or allocate dynamically based
+    on load. If ncpus is None and maxload is None, then it will use the number of CPUs
+    returned by multiprocessing; if ncpus is not None, it will use the specified number;
+    if maxload is not None, it will allocate dynamically.
+    
+    Example 1:
+        def f(x):
+            return x*x
+        
+        results = parallelize(func=f, iterarg=[1,2,3], ncpus=3)
+    
+    Example 2:
+        import pylab as pl
+        def myfunc(x, y, i):
+            xy = [x+i*pl.rand(1000), y+i*pl.randn(1000)]
+            return xy
+        n = 5
+        xy = parallelize(myfunc, kwargs={'x':3, 'y':8}, iterarg=n, itervar='i', maxload=0.8)
```

Comment:
Is there an example to do something like this (this doesn't work):

```python
def f(x,y):
    return x*y
results = sc.parallelize(func=f,iterarg=[(1,2),(2,3),(3,4)],ncpus=3) # fails
results = sc.parallelize(func=f,iterarg=[{'x':1,'y':2},{'x':2,'y':3},{'x':3,'y':4}],ncpus=3) # fails
```

Doesn't need to be exactly that syntax, maybe it's fine with the existing functionality, but could be useful to see that as a worked example

Reply:
Ah yes that should definitely work! Should be fixed now.